File: 0118_pascals_triangle.rb
require 'pry'

def generate(num_rows)
    i = 1
    ans = []
    prev = nil
    until i > num_rows
        row = Array.new(i) {1}
        if i > 2
            row = mod_middle(row, prev)
        end
        ans << row
        prev = ans[i - 1]
        i += 1
    end
    ans
end

def mod_middle(row, prev)
    j = 1
    until j == row.length
        row[j] = prev[j] + prev[j - 1]
        j += 1
    end
    row
end

generate(5)

File: 0118_pascals_triangle.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        #edge cases
        if numRows == 0:
            return []
        elif numRows == 1:
            return [[1]]
        elif numRows == 2:
            return [[1], [1, 1]]

        counter = 3
        ans = [[1], [1, 1]]
        prev = [1, 1]

        while (len(ans) < numRows):
            row = [1] * counter
            for i in range(1, len(row) - 1):
                row[i] = prev[i - 1] + prev[i]
            ans.append(row)
            prev = row
            counter += 1
        return ans
